posts/java-enums-with-extra-visitor-please.md

```diff
@@ -0,0 +1,203 @@
+Using enums can be tricky when it comes to making decisions according to its values. As the enum evolves, each part of the code where it has been used as a condition must be checked. If many decisions have been based on an enum's values, maintaining the code base can be a nightmare. A single missed verification can lead to corruption of the entire system.
+
+Is there a way to reduce the impact of enums' impact on the code base?
+
+# The switch-case approach
+
+Consider the `AssetClass` enum which represents distinct groups of tradable commodities:
+
+```java
+public enum AssetClass {
+    METAL,
+    ENERGY,
+    AGRICULTURAL,
+}
+```
+
+This enum can be used to decide which strategy, mapper or any other behavior to use according to the value of `AssetClass`. A general example of this is given in the following code:
+
+```java
+public AutomatedTradingStrategy getAutomatedTradingStrategy(AssetClass assetClass) {
+    switch (assetClass) {
+        case METAL:return new HedgingStrategy();
+        case ENERGY: return new SwingTradingStrategy();
+        case AGRICULTURAL:
+        default: return DayTradingStrategy();
+    }
+}
+```
+
+The switch-case statement is probably the most straightforward way to do this. However, it has several flaws.
+
+The method `getAutomatedTradingStrategy()` returns a behavior according to the value of `AssetClass`. Defining a default behavior then becomes mandatory, even if in this example all the values of `AssetClass` enum are handled. We can do this by either returning a default implementation of `AutomatedTradingStrategy`, null, or throwing an exception.
+
+Even though each enum value should have a handler adding a default handler here ensures that there is some way to handle an arbitrary `AssetClass`.  This is helpful when an enum value has been added but is missed in other parts of the codebase.
+
+The last problem is probably the least obvious. Using the switch-case statement generates a strong coupling between the business logic and the enum's values, breaking the (open-closed principle)[https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle].
+Yet, the switch-case statement has no interest in knowing if the asset class is an enum, object, or anything else. Only the semantic matters.
+For instance, metals could be split into two sub assets: base metals and precious metals. Any already existing code based on `AssetClass.METAL` will have to be reworked to take this change into account. No business value has been added where the rework was necessary while it exposed a working implementation to the risk of regressions.
+
+
+# Visitor pattern to the rescue
+
+How can we break this coupling while offering the ability to contextualize the decision making to the enum's values? The answer is in the title: let's use the (Visitor pattern)[https://en.wikipedia.org/wiki/Visitor_pattern].
+
+First off, we need to create an interface which will be used as a contract between the enum and the code relying on its values.
```

Comment:
First, we need to create an interface which will be used as a contract between the enum and the code relying on its values.